test/test_public_bindings.py

```diff
@@ -272,6 +278,68 @@ def test_no_new_bindings(self):
+        and there is no name in `__module__` or the object itself that starts with “_”.
+        Each public package should either:
+        - (preferred) Define `__all__` and all callables and classes in there must have their
+         `__module__` start with the current submodule's path. Things not in `__all__` should
+          NOT have their `__module__` start with the current submodule.
+        - (for simple python-only modules) Not define `__all__` and all the elements in `dir(submod)` must have their
+          `__module__` that start with the current submodule.
+        '''
+        failure_list = []
+        with open(os.path.join(sys.path[0], 'allowlist_for_publicAPI.json')) as json_file:
+            allow_dict = json.load(json_file)
+
+        def test_module(modname):
+            split_strs = modname.split('.')
+            mod = sys.modules.get(modname)
+            for elem in split_strs:
+                if elem.startswith("_"):
+                    return
+
+            def add_to_failure_list_if_not_in_allow_dict(modname, elem, elem_module):
+                if modname in allow_dict and elem in allow_dict[modname]:
+                    return
+                failure_list.append((modname, elem, elem_module))
+
+            # verifies that each public API has the correct module name and naming semantics
+            def looks_public_or_not(elem, modname, mod, is_public=True):
+                obj = getattr(mod, elem)
+                if not (isinstance(obj, Callable) or inspect.isclass(obj)):
+                    return
+                elem_module = getattr(obj, '__module__', None)
+                elem_modname_starts_with_mod = elem_module is not None and elem_module.startswith(modname)
+                # elem's name must NOT begin with an `_` and it's module name
+                # SHOULD start with it's current module since it's a public API
+                looks_public = not elem.startswith('_') and elem_modname_starts_with_mod
+                if is_public == looks_public:
+                    add_to_failure_list_if_not_in_allow_dict(modname, elem, elem_module)
+
+            if hasattr(modname, '__all__'):
+                public_api = mod.__all__
+                all_api = dir(modname)
+                for elem in all_api:
+                    looks_public_or_not(elem, modname, is_public=elem in public_api)
+
+            else:
+                all_api = dir(mod)
+                for elem in all_api:
+                    if not elem.startswith('_'):
+                        looks_public_or_not(elem, modname, mod, is_public=True)
+
+        for _, modname, ispkg in pkgutil.walk_packages(path=torch.__path__, prefix=torch.__name__ + '.'):
+            test_module(modname)
+
+        test_module('torch')
+        msg = "Following new APIs ( displayed in the form (module, element, element module) )" \
+              " were added that do not meet our guidelines for public API" \
```

Comment:
We can link to https://docs.google.com/document/d/10yx2-4gs0gTMOimVS403MnoAWkqitS8TUHX73PN8EjE/edit?pli=1# here. 
Could you also leave a TODO comment in the code to move that to the wiki.